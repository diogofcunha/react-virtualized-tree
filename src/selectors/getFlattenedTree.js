export const isNodeExpanded = node => node.state && node.state.expanded;
export const nodeHasChildren = node => node.children && node.children.length;

export const getFlattenedTree = (nodes, parents = []) =>
  nodes.reduce((flattenedTree, node) => {
    const deepness = parents.length;
    const nodeWithHelpers = {...node, deepness, parents};

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      return [...flattenedTree, nodeWithHelpers];
    }

    return [...flattenedTree, nodeWithHelpers, ...getFlattenedTree(node.children, [...parents, node.id])];
  }, []);

export const getFlattenedTreePaths = (nodes, parents = []) => {
  const paths = [];

  for (const node of nodes) {
    const {id} = node;

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      paths.push(parents.concat(id));
    } else {
      paths.push(parents.concat(id));
      paths.push(...getFlattenedTreePaths(node.children, [...parents, id]));
    }
  }

  return paths;
};

export const doesChangeAffectFlattenedTree = (previousNode, nextNode) => {
  return isNodeExpanded(previousNode) !== isNodeExpanded(nextNode);
};
