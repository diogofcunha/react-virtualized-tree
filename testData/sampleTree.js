export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
