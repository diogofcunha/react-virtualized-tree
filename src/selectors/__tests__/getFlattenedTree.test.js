import {getFlattenedTree, getFlattenedTreePaths} from '../getFlattenedTree';
import {Nodes} from '../../../testData/sampleTree';

describe('getFlattenedTree', () => {
  it('should match snapshot', () => {
    expect(getFlattenedTree(Nodes)).toMatchSnapshot();
  });
});

test('getFlattenedTreePaths, should match snapshot', () => {
  expect(getFlattenedTreePaths(Nodes)).toMatchSnapshot();
});
