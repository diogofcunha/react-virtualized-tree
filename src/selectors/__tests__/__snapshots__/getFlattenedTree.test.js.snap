// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFlattenedTree should match snapshot 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "id": "c-3",
                "name": "Leaf 3 Child",
                "state": Object {},
              },
            ],
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "expanded": false,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "deepness": 0,
    "id": 0,
    "name": "Leaf 1",
    "parents": Array [],
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": "c-3",
            "name": "Leaf 3 Child",
            "state": Object {},
          },
        ],
        "id": 3,
        "name": "Leaf 3",
        "state": Object {
          "deletable": true,
          "expanded": false,
          "favorite": true,
        },
      },
      Object {
        "id": 4,
        "name": "Leaf 4",
      },
    ],
    "deepness": 1,
    "id": 2,
    "name": "Leaf 2",
    "parents": Array [
      0,
    ],
    "state": Object {
      "deletable": true,
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "id": "c-3",
        "name": "Leaf 3 Child",
        "state": Object {},
      },
    ],
    "deepness": 2,
    "id": 3,
    "name": "Leaf 3",
    "parents": Array [
      0,
      2,
    ],
    "state": Object {
      "deletable": true,
      "expanded": false,
      "favorite": true,
    },
  },
  Object {
    "deepness": 2,
    "id": 4,
    "name": "Leaf 4",
    "parents": Array [
      0,
      2,
    ],
  },
  Object {
    "deepness": 1,
    "id": 5,
    "name": "Leaf 5",
    "parents": Array [
      0,
    ],
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "deepness": 0,
    "id": 1,
    "name": "Leaf 6",
    "parents": Array [],
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "deepness": 0,
    "id": "z",
    "name": "Leaf z",
    "parents": Array [],
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`getFlattenedTreePaths, should match snapshot 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    2,
    3,
  ],
  Array [
    0,
    2,
    4,
  ],
  Array [
    0,
    5,
  ],
  Array [
    1,
  ],
  Array [
    "z",
  ],
]
`;
